File: cpp/baseline.cpp
#define _CRT_SECURE_NO_WARNINGS 1 

#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <cmath>

using namespace std;

#define MAXN 5000
#define MAXC 50
#define MAXB 50
#define MAXW 100
#define MAXH 10

typedef struct item{
	int ind;
	int cat;
	int brand;
	int c;
} item;

vector<item> items;
vector<item> cat2items[MAXC];

bool compare_items(const item &a, const item &b) {
	return a.c > b.c || a.c == b.c && a.brand < b.brand;
}

bool compare_items_by_brand(const item &a, const item &b) {
	return a.brand < b.brand || a.brand == b.brand && a.c > b.c;
}


typedef struct rectangle {
	int w;
	int h;
} rectangle;

double cat_score[MAXC + 1][MAXH + 1][MAXW + 1] = { 0 };
vector<item> cat_score_items[MAXC + 1][MAXH + 1][MAXW + 1];
double c[MAXC + 1][MAXW + 1] = { 0 };
rectangle c_best[MAXC + 1][MAXW + 1];

item sh[MAXH][MAXW];
double score_matrix[MAXH][MAXW] = { 0 };


double calculate_brand_score(item a[MAXH][MAXW], int h, int w) {
	for (int i = 0; i < h; i++) {
		for (int j = 0; j < w; j++) {
			score_matrix[i][j] = a[i][j].c;
		}
	}
	for (int i = 0; i < h; i++) {
		vector<int> state;
		for (int j = i; j < h; j++) {
			int left = 0;
				
			for (int k = 0; k <= w; k++) {
				if (k != w) {
					if (i == j) {
						state.push_back(a[i][k].brand);
					}
					else {
						if (state[k] != a[j][k].brand) {
							state[k] = -1;
						}
					}
				}
				if (k > 0) {
					if (state[k - 1] != -1 && (k == w || state[k] != state[k - 1])) {
						double a_score = (1 + log(1.0 * (j - i + 1) * (k - left)) / log(2));
						for (int u = i; u <= j; u++) {
							for (int q = left; q < k; q++) {
								score_matrix[u][q] = max(score_matrix[u][q], 1.0 * a[u][q].c * a_score);
							}
						}
					}
					if (k != w) {
						if (state[k] == -1 || state[k] != state[k - 1]) {
							left = k;
						}
					}
				}
			}
		}
	}
	double res = 0;
	for (int i = 0; i < h; i++) {
		for (int j = 0; j < w; j++) {
			res += score_matrix[i][j];
		}
	}
	return res;
}

double calculate_cat_score(item a[MAXH][MAXW], int h, int w, int n_cat, int d0) {
	vector<int> min_h, max_h, min_w, max_w, cnt;
	for (int i = 0; i < n_cat; i++) {
		min_h.push_back(h + 1);
		max_h.push_back(-1);
		min_w.push_back(w + 1);
		max_w.push_back(-1);
		cnt.push_back(0);
	}
	for (int i = 0; i < h; i++) {
		for (int j = 0; j < w; j++) {
			if (a[i][j].cat != -1) {
				min_h[a[i][j].cat - 1] = min(min_h[a[i][j].cat - 1], i);
				max_h[a[i][j].cat - 1] = max(max_h[a[i][j].cat - 1], i);
				min_w[a[i][j].cat - 1] = min(min_w[a[i][j].cat - 1], j);
				max_w[a[i][j].cat - 1] = max(max_w[a[i][j].cat - 1], j);
				cnt[a[i][j].cat - 1] += 1;
			}
		}
	}
	double score = 0;
	for (int i = 0; i < n_cat; i++) {
		if (cnt[i] > 0) {
			if (cnt[i] != (max_h[i] - min_h[i] + 1) * (max_w[i] - min_w[i] + 1)) {
				return -1;
			}
			score += 1.0 * d0 * sqrt(1.0 * (cnt[i]) / (w * h));
		}
	}
	return score;
}

double calculate_final_score(item a[MAXH][MAXW], int h, int w, int n_cat, int d0) {
	double cat_score = calculate_cat_score(a, h, w, n_cat, d0);
	if (cat_score == -1) {
		return -1;
	}
	double brand_score = calculate_brand_score(a, h, w);
	return cat_score + brand_score;
}


void run() {

	std::cout.precision(5);
	std::cout.setf(ios::fixed);

	int n, k, m, h, w, d0, ti, bi, ci;
	cin >> n >> k >> m >> h >> w >> d0;
	for (int i = 0; i < k; i++) {
		cat2items[i] = vector<item>();
	}
	for (int i = 0; i < n; i++) {
		cin >> ti >> bi >> ci;
		cat2items[ti - 1].push_back({ (i + 1), ti, bi, ci });
	}

	
	/*
	for each category sort items by cost in descending order(and secondary sort by brand for reproducibility)
	see compare_items function at the top of file
	*/
	for (int i = 0; i < k; i++) {
		sort(cat2items[i].begin(), cat2items[i].end(), compare_items);
	}

	/*
	calculate for each triple (cat - category, h - rectangle height, w - rectangle width):
		cat_score[cat][h][w] - estimate of score using greedy filling approach
		cat_score_items[cat][h][w] - ordered list of items for shelf filling
	*/
	for (int i = 0; i < k; i++) {
		for (int hh = 1; hh <= h; hh++) {
			for (int ww = 1; ww <= w; ww++) {
				double simple_brand_score = 0;
				int n_real = min((int)cat2items[i].size(), ww * hh);
				for (int j = 0; j < n_real; j++) {
					simple_brand_score += cat2items[i][j].c;
				}
				cat_score_items[i][hh][ww].clear();
				vector<item> tmp;
				for (int j = 0; j < n_real; j++) {
					tmp.push_back(cat2items[i][j]);
				}
				// group(sort) choosed items by brand (and secondary sort by cost for reproducibility)
				// (see compare_items_by_brand function at the top of file)
				sort(tmp.begin(), tmp.end(), compare_items_by_brand);
				for (int j = 0; j < n_real; j++) {
					cat_score_items[i][hh][ww].push_back(tmp[j]);
				}
				cat_score[i][hh][ww] = simple_brand_score + 1.0 * d0 * sqrt(1.0 * (n_real) / (w * h));
			}
		}
	}

	/*
	DP(Dynamic Programming) part to find optimal rectangle size for each category
		c[i][j] - best estimate of score for shelf size = (h x j) if we used only categories 1..i
		c_best[i][j] - for shelf size = (h x j) stores best rectangle(w_cat, h_cat) for i-th category
		(is necessary for fast answer recovering)
	*/
	for (int i = 0; i <= k; i++) {
		for (int j = 0; j <= w; j++) {
			c[i][j] = 0;
			c_best[i][j] = { 0, 0 };
		}
	}
	for (int i = 0; i < k; i++) {
		for (int j = 0; j <= w; j++) {
			int n_items_in_cat_i = cat2items[i].size();
			// brute force all possible rectangle sizes for i-th category
			for (int w_cat = 0; w_cat <= min(n_items_in_cat_i, w - j); w_cat++) {
				for (int h_cat = 0; h_cat * w_cat <= n_items_in_cat_i && h_cat <= h; h_cat++) {
					double new_score = c[i][j] + cat_score[i][h_cat][w_cat];
					if (c[i + 1][j + w_cat] < new_score) {
						c[i + 1][j + w_cat] = new_score;
						c_best[i + 1][j + w_cat] = { w_cat, h_cat };
					}
				}
			}
		}
	}

	/*
	Answer recovering part using c_best matrix
	*/
	for (int i = 0; i < h; i++) {
		for (int j = 0; j < w; j++) {
			sh[i][j] = { 0, -1, -1, 0 };
		}
	}
	int i = k;
	int j = w;
	while (i > 0) {
		rectangle ij_best = c_best[i][j];
		int cc = 0;
		for (int u = 0; u < ij_best.h; u++) {
			for (int q = j - ij_best.w; q < j; q++) {
				sh[u][q] = cat_score_items[i - 1][ij_best.h][ij_best.w][cc];
				cc += 1;
			}
		}
		i -= 1;
		j -= ij_best.w;
	}

	for (int i = 0; i < h; i++) {
		for (int j = 0; j < w; j++) {
			std::cout << sh[i][j].ind << " ";
		}
		std::cout << endl;
	}
}

int main(int argc, char *argv[])
{
	run();
	return 0;
}

File: py/baseline.py
import sys
from typing import NamedTuple, Tuple, List

# from base import Item, Rectangle, create_empty_matrix
# ----------copy from base.py---------- #
class Item(NamedTuple):
    ind: int
    cat: int
    brand: int
    c: int


class Rectangle(NamedTuple):
    w: int
    h: int


def create_empty_matrix(h: int, w: int) -> List[List[Item]]:
    return [[Item(0, -1, -1, 0) for _ in range(w)] for _ in range(h)]
# ----------copy from base.py---------- #


def run():
    n, k, m, h, w, d0, items = read_input(sys.stdin)

    cat2items: List[List[Item]] = []
    for i in range(k):
        cat2items.append([])

    for item in items:
        cat2items[item.cat - 1].append(item)

    '''
    for each category sort items by cost in descending order (and secondary sort by brand for reproducibility)
    '''
    for i in range(k):
        cat2items[i] = sorted(cat2items[i], key=lambda x: (-x.c, x.brand))

    '''
    calculate for each triple (cat - category, h - rectangle height, w - rectangle width):
        cat_score[cat][h][w] - estimate of score using greedy filling approach
        cat_score_items[cat][h][w] - ordered list of items for shelf filling 
    '''
    cat_score: List[List[List[float]]] = []
    cat_score_items: List[List[List[List[Item]]]] = []
    for i in range(k):
        cat_score_i = []
        cat_score_i_items = []
        for hh in range(h + 1):
            cat_score_i.append([0] * (w + 1))
            cat_score_i_items.append([None] * (w + 1))

        for hh in range(1, h + 1):
            for ww in range(1, w + 1):
                simple_brand_score = 0
                n_real = min(len(cat2items[i]), ww * hh)
                for j in range(n_real):
                    simple_brand_score += cat2items[i][j].c

                cat_score_i[hh][ww] = simple_brand_score + d0 * (n_real / (w * h))**0.5
                # group(sort) choosed items by brand (and secondary sort by cost for reproducibility)
                cat_score_i_items[hh][ww] = sorted(cat2items[i][:n_real].copy(), key=lambda x: (x.brand, -x.c))
        cat_score.append(cat_score_i)
        cat_score_items.append(cat_score_i_items)

    '''
    DP(Dynamic Programming) part to find optimal rectangle size for each category
        c[i][j] - best estimate of score for shelf size = (h x j) if we used only categories 1..i
        c_best[i][j] - for shelf size = (h x j) stores best rectangle (w_cat, h_cat) for i-th category 
                       (is necessary for fast answer recovering)
    '''
    c: List[List[float]] = []
    c_best: List[List[Rectangle]] = []
    for i in range(k + 1):
        c.append([0] * (w + 1))
        c_best.append([Rectangle(0, 0)] * (w + 1))

    for i in range(k):
        for j in range(w + 1):
            n_items_in_cat_i = len(cat2items[i])
            # brute force all possible rectangle sizes for i-th category
            for w_cat in range(min(n_items_in_cat_i, w - j) + 1):
                for h_cat in range(h + 1):
                    if h_cat * w_cat > n_items_in_cat_i:
                        break
                    new_score = c[i][j] + cat_score[i][h_cat][w_cat]
                    if c[i + 1][j + w_cat] < new_score:
                        c[i + 1][j + w_cat] = new_score
                        c_best[i + 1][j + w_cat] = Rectangle(w_cat, h_cat)

    '''
    Answer recovering part using c_best matrix
    '''
    res = create_empty_matrix(h, w)
    i, j = k, w
    while i > 0:
        rectangle = c_best[i][j]
        cc = 0
        for u in range(rectangle.h):
            for q in range(j - rectangle.w, j):
                res[u][q] = cat_score_items[i - 1][rectangle.h][rectangle.w][cc]
                cc += 1
        i -= 1
        j -= rectangle.w

    for i in range(h):
        print(' '.join([str(e.ind) for e in res[i]]))


def read_input(fin) -> Tuple[int, int, int, int, int, int, List[Item]]:
    n, k, m, h, w, d0 = map(int, fin.readline().strip().split())
    items = []
    for i in range(n):
        ti, bi, ci = map(int, fin.readline().strip().split())
        items.append(Item((i + 1), ti, bi, ci))
    return n, k, m, h, w, d0, items


if __name__ == '__main__':
    run()
